Allow RMG to handle a specific species as a third body collider

alongd opened the pull request:
Previously RMG could not handle reactions of the sort `A + B (+C) = D (+C)`. Moreover, reactions like `H2 + Ar <=> H + H + Ar` were only handled correctly if the kinetics is of Arrhenius type (like in the BurkeH2O2 libraries), but if the kinetics is in any PDep form, RMG would consider it as `H2 + Ar (+M) <=> H + H + Ar (+M)`, wrongly accounting for an additional third body.

**Usage**
Simple add a specific third body collider to your library reaction as you would intuitively (i.e., no need for a SpecificCollider keyword).
For example (and this rxn is the motivation for this PR):
```
entry(
    index = 226,
    label = "SO2 + O (+N2) <=> SO3 (+N2)",
    degeneracy = 1,
    kinetics = Troe(
        arrheniusHigh = Arrhenius(A=(3.7e+11, 'cm^6/(mol^2*s)'), n=0, Ea=(1689, 'cal/mol'), T0=(1, 'K')),
        arrheniusLow = Arrhenius(A=(2.9e+27, 'cm^9/(mol^3*s)'), n=-3.58, Ea=(5206, 'cal/mol'), T0=(1, 'K')),
        alpha=0.43, T3=(371, 'K'), T1=(7442, 'K'), efficiencies={}),
    shortDesc = u"""[Marshall2005],[Marshall2006]""",
)
```

Another example w/o parenthesis, which RMG should also handle correctly now, is:
```
entry(
    index = 18,
    label = "HNCN + N2 <=> H + NCN +N2",
    degeneracy = 1,
    kinetics = ThirdBody(
        arrheniusLow = Arrhenius(A=(1.79e+28, 'cm^3/(mol*s)'), n=-3.44, Ea=(64502, 'cal/mol'), T0 = (1, 'K'), Tmin=(2000, 'K'), Tmax=(4000, 'K'))),
    shortDesc = u"""[Lin2000a]""",
)
```

Currently we only allow one specific collider per reaction (As far as I can see, I don't think more than one would be necessary, and probably also isn't implemented in Chemkin/Cantera, although I haven't checked). E.g., `A + B (+N2) (+Ar) <=> C (+N2) (+Ar)` isn't valid in RMG, and will throw [an appropriate error](https://github.com/ReactionMechanismGenerator/RMG-Py/commit/7d85bfcd09ce0de7eaf877676bbd6791cd94890e#diff-19d238ba2c9b6d51bc800c3dd7a1278eR278).

The implementation of this feature in the solver (simple.pyx) is through the [Peff calculation](https://github.com/ReactionMechanismGenerator/RMG-Py/commit/877c70cd63a96e183dc0eac44068f79309784768#diff-0d43fdf5d1b8f1dbecaa89c49a4afadbR192).

Two tests were added:
- testReadSpecificCollider in chemkinTest.py
- testSpecificColliderModel in simpleTest.py

Lastly, this PR also allows RMG to consider specificColliders species with parenthesis in their name such as `N2(4)` or `CH2(S)` or `CH2(S)(15)`, as was recently fixed in [Cantera](https://github.com/Cantera/cantera/pull/460) and will be added to the next Chemkin release.

Addresses #1070

alongd commented:
This specificCollider PR currently fails on the new testSpecificColliderModel test.
I'm still looking into it, but would love another set of eyes helping me figure out whether an additional change to the solver should be made :hushed: :thought_balloon:

rwest commented:
I don't have time to look properly at this this week, but I did have a question about your second example above:
```
entry(
    index = 18,
    label = "HNCN + N2 <=> H + NCN +N2",
    degeneracy = 1,
    kinetics = ThirdBody(
        arrheniusLow = Arrhenius(A=(1.79e+28, 'cm^3/(mol*s)'), n=-3.44, Ea=(64502, 'cal/mol'), T0 = (1, 'K'), Tmin=(2000, 'K'), Tmax=(4000, 'K'))),
    shortDesc = u"""[Lin2000a]""",
)
```

What is this meant to represent? 

If it's just _r˲_ = _k_[HNCN][N₂] then wouldn't 
`Arrhenius(A=(1.79e+28, 'cm^3/(mol*s)'), n=-3.44, Ea=(64502, 'cal/mol'))`
fully describe it, rather than 
`ThirdBody(arrheniusLow=Arrhenius(A=(1.79e+28, 'cm^3/(mol*s)'), n=-3.44, Ea=(64502, 'cal/mol')))` ?

Perhaps you could say how it would look in CHEMKIN syntax?  Or how would you describe the kinetics?

alongd commented:
I agree, there's no advantage in writing a reaction with a specificCollider in a ThirdBody format, i.e.:
```
entry(
    index = 1,
    label = "H + O2 + CO2 <=> O + OH + CO2",
    degeneracy = 1,
    kinetics = ThirdBody(
        arrheniusLow = Arrhenius(A=(1.79e+28, 'cm^6/(mol^2*s)'), n=-3.44, Ea=(64502, 'cal/mol'), T0 = (1, 'K'), Tmin=(2000, 'K'), Tmax=(4000, 'K'))),
)
```
and it is indeed much simpler (and recommended?) to  write it as a normal Arrhenius, as pointed out by @rwest. 

However, if a user happens to use the ThirdBody kinetics format for a library reaction with a specific collider, we would like RMG to handle that correctly. Before this fix, the reaction in the example above would have the following wrong Chemkin representation (with two colliders):

```
! Reaction index: Chemkin #2; RMG #2
! Library reaction: FFCM1(-)
! Flux pairs: CO2(6), CO2(6); O2(1), O(7); H(3), OH(4); 
H(3)+O2(1)+CO2(6)+M=OH(4)+O(7)+CO2(6)+M             1.790e+34 -3.440    64.502   
```

But now RMG will pick that up, add a comment that a specificCollider was identified, and output the kinetics as if a normal Arrhenius kinetics reaction was inputted:
```
! Reaction index: Chemkin #2; RMG #2
! Library reaction: FFCM1(-)
! Specific third body collider: CO2
! Flux pairs: O2(1), O(6); H(3), OH(4); 
H(3)+O2(1)+CO2(7)=OH(4)+O(6)+CO2(7)                 1.790e+28 -3.440    64.502   
```

alongd commented:
Added kinetics documentation. Also addresses [RMG-database#191](https://github.com/ReactionMechanismGenerator/RMG-database/issues/191)

alongd commented:
Couldn't figure out how to add relative references to the Classes in modification.rst pointing outside the `users` folder in the documentation (so added absolute).
Any ideas?

rwest commented:
If the sole aim is to protect against user error, I would suggest stopping with a helpful error message, rather than trying to guess what they intended. If they made a mistake, then we may guess incorrectly. 

(Also, rather unlikely, but if someone actually did want to input "H + O2 + CO2 +M <=> O + OH + CO2 +M"  or "A + CO2 + M <=> B + CO2 + M" there would be no way to do so in the new format?)

alongd commented:
@rwest: Will do. Thanks for that.
Let's see that I got it right: a reaction with a third body collider in parenthesis, whether specific or not, should only be applied to kinetics formats that specify High and Low rates with different corresponding orders (units).
In other words, we should only allow the parenthesis form `(+M)` or `(+specificSpecies)` in Troe and Lindemann, right?

alongd commented:
I see that Chebyshev also gets a (+M), meaning a specificCollider should be allowed there as well?

mjohnson541 commented on documentation/source/users/rmg/database/modification.rst:
```diff
@@ -73,6 +72,167 @@ There are several places in the RMG-database and RMG-Py source code where reacti
+
+**MultiArrhenius**
+(see `MultiArrhenius Class <http://reactionmechanismgenerator.github.io/RMG-Py/reference/kinetics/multiarrhenius.html>`_ for details)::
+
```
Can you specify that MultiArrhenius is the sum of multiple Arrhenius expressions (as opposed to being one Arrhenius over one Trange and one Arrhenius over another Trange)

mjohnson541 commented on documentation/source/users/rmg/database/modification.rst:
```diff
@@ -73,6 +72,167 @@ There are several places in the RMG-database and RMG-Py source code where reacti
+	        kunits='cm^3/(mol*s)', Tmin=(300, 'K'), Tmax=(3000, 'K'), Pmin=(0.0013156, 'atm'), Pmax=(131.56, 'atm')))
+
+
+The ``Troe``, ``Lindemann``, and ``Chebyshev`` pressure dependent formats could also be defined with a specific collider if needed, for example::
```
Can you add a title so this doesn't get passed over as part of the Chebyshev section?

rwest commented:
I don't know how chemkin would interpret `(+N2)`  (meaning a specificCollider) on a CHEB or a PLOG, both of which are defined in terms of pressure (Atm) rather than species concentrations. Perhaps it would figure out the partial pressure of N2.  I have never seen such an expression "in the wild".

Now that we parse things like `(+N2)` in our RMG reaction definitions, I wonder if we should parse (ad require) `(+M)` to signify pressure-dependent reactions, and `+M` to signify third-body reactions (low pressure limit), like Chemkin and Cantera.  Consequences would have to be thought through, but it might reduce confusion for people translating between RMG and Chemkin or Cantera formats. 

At the moment the chemkin reactions `A (+M) = B + C (+M)`  and `A = B + C` and `A +M = B + C +M` are all written as `A = B + C` in RMG (but with different kinetics attributes).

Perhaps not for this pull request, but might be worth thinking about.

alongd commented:
While Chemkin's pre-processor failed on a PLOG reaction with a specific collider:
```
  1. CH3(17)+CH2NH2(91)(+Ar)=NCC(1)(+Ar)            1.00E+00    0.0        0.0
      Error...PLOG and other pressure-depency mutually exclusive...
      Rate coefficients at P=4.93E-01(atm)           9.381E+54  -12.3       22.7
      Rate coefficients at P=7.72E-01(atm)           4.176E+52  -11.6       21.8
      Rate coefficients at P=1.21E+00(atm)           1.011E+50  -10.8       20.6
      Rate coefficients at P=1.89E+00(atm)           1.393E+47   -9.9       19.2
      Rate coefficients at P=2.96E+00(atm)           1.186E+44   -9.0       17.7
      Error...no pres-dependency option defined...
```
it did allow it in a CHEB reaction (also, RMG outputs CHEB with `(+M)`, but not PLOG).
I'm not sure why CHEB gets a different treatment in this respect (but at least it's consistent).

The current state of this branch is to allow a specific collider only for reactions that get a `(+M)`: Troe, Lindemann, and Chebyshev.

The Chebyshev case is indeed strange. Generally, if a specific collider is present, RMG calculates the rate at the partial pressure of the collider species alone. For Chebyshev this could be a problem if it isn't defined for this specific (low) pressure.

I tried running a CHEB format with a specific collider again in Chemkin, this time using a minor species (not Ar). Chemkin failed right away since the intermediate species' concentration was 0 in the beginning of the simulation.

I think we shouldn't allow specific colliders in Chebyshev, although using one of the bath gases as a specific collider could make some sense.

alongd commented:
Thanks @mjohnson541!
I addressed all comments, the fixes for library.py and family.py were already squashed.
We now also throw an error if a specificCollider is given for Chebyshev, following @rwest's and my comments above (unless there are any objections?)
Let me know if everything's OK, and I'll organize and squash.

alongd commented:
BTW, what's up with Travis lately?
